NeuroID/Utilities/NIDConfigService.swift

```diff
@@ -6,56 +6,95 @@ import Alamofire
+    static var NID_CONFIG_URL = "https://scripts.neuro-id.com/mobile/"
     
-    static var cacheSetWithRemote = false
+    var networkService: NIDNetworkServiceProtocol
+
+    var cacheSetWithRemote = false
+    var cacheCreationTime: Date = .init()
+
+    public var configCache: ConfigResponseData = .init()
     
-    public static var nidConfigCache: ResponseData = .init()
-        
-    static var nidURL = "https://scripts.neuro-id.com/mobile/"
+    init(networkService: NIDNetworkServiceProtocol = NeuroID.networkService) {
+        self.networkService = networkService
+    }
     
-    init(completion: @escaping (Bool) -> Void) {
-        if NeuroID.clientKey == nil || NeuroID.clientKey == "" {
-            NIDLog.e("Missing Client Key. Config Service not started")
-            completion(true)
+    func retrieveConfig(completion: @escaping () -> Void) {
+        if !NeuroID.verifyClientKeyExists() {
+            cacheSetWithRemote = false
+            completion()
             return
         }
-                     
-        let config_url = NIDConfigService.nidURL + NeuroID.clientKey! + ".json"
-        AF.request(config_url, method: .get).responseDecodable(of: ResponseData.self) { response in
+        
+        let config_url = NIDConfigService.NID_CONFIG_URL + NeuroID.clientKey! + ".json"
+        networkService.getRequest(
+            url: URL(string: config_url)!,
+            responseDecodableType: ConfigResponseData.self
+        ) { response in
             switch response.result {
             case .success(let responseData):
                 self.setCache(responseData)
-                NIDLog.d("Retrieved config log")
-                NIDConfigService.cacheSetWithRemote = true
-                completion(true)
+                NIDLog.d("Retrieved remote config")
+                
+                self.cacheSetWithRemote = true
+                self.cacheCreationTime = .init()
+                completion()
             case .failure(let error):
                 NIDLog.e("Failed to retrieve NID Config \(error)")
-                NIDConfigService.nidConfigCache = ResponseData()
-                completion(true)
+                self.configCache = ConfigResponseData()
+                self.cacheSetWithRemote = false
+                completion()
             }
         }
     }
     
-    func setCache(_ responseData: ResponseData) {
-        NIDConfigService.nidConfigCache = responseData
+    func setCache(_ newCache: ConfigResponseData) {
+        configCache = newCache
+    }
+    
+    func expiredCache() -> Bool {
+        if !cacheSetWithRemote {
+            return true
+        }
+        // 5 min is the default, can be updated later
+        let TTLTime = Calendar.current.date(byAdding: .minute, value: -5, to: Date())!
+        
+        return cacheCreationTime < TTLTime
+    }
+    
+    func retrieveOrRefreshCache(completion: @escaping () -> Void) {
+        if expiredCache() {
+            retrieveConfig {
```

Comment:
We should consider sending a `SET_VARIABLE` event when the config is refreshed with the config version, so that we can later know what config was applied during collection

Reply:
That's a good idea, we don't currently have a concept of a config "version" but its something we can add a log saying the config was refreshed and then add a version to the config at a later point

Reply:
We should at least encode the current config into the event. Ive spent way too much time debugging web sessions where the config didnt match what we _THOUGHT_ the config was only to find out it wasnt, and that was causing the bug.